File: app/src/main/java/com/codebosses/comical/fragments/FragmentSearch.java
package com.codebosses.comical.fragments;

import android.content.Intent;
import android.os.Bundle;

import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.codebosses.comical.R;
import com.codebosses.comical.activities.ComicDetailActivity;
import com.codebosses.comical.adapters.SearchComicAdapter;
import com.codebosses.comical.databinding.FragmentSearchBinding;
import com.codebosses.comical.endpoints.EndpointKeys;
import com.codebosses.comical.fragments.base.BaseFragment;
import com.codebosses.comical.pojo.Comic;
import com.codebosses.comical.pojo.event_bus.EventBusAdapterClick;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;

public class FragmentSearch extends BaseFragment {

    private static FragmentSearch fragmentSearch;

    //    Android fields....
    private FragmentSearchBinding searchBinding;

    //    Adapter fields....
    private SearchComicAdapter searchComicAdapter;
    private List<Comic> comicsList = new ArrayList<>();

    public static FragmentSearch getInstance() {
        if (fragmentSearch == null) {
            fragmentSearch = new FragmentSearch();
        }
        return fragmentSearch;
    }

    public FragmentSearch() {

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        searchBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_search, container, false);

        searchComicAdapter = new SearchComicAdapter(getActivity(), comicsList);
        searchBinding.recyclerViewSearch.setLayoutManager(new LinearLayoutManager(getActivity()));
        searchBinding.recyclerViewSearch.setAdapter(searchComicAdapter);

        return searchBinding.getRoot();
    }

    @Override
    public void onStart() {
        super.onStart();
        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this);
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        if (EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this);
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onSearchComicClick(EventBusAdapterClick eventBusAdapterClick) {
        if (eventBusAdapterClick.getClickType().equals(EndpointKeys.SEARCH_COMIC_CLICK)) {
            Intent intent = new Intent(getActivity(), ComicDetailActivity.class);
//            intent.putExtra(EndpointKeys.COMIC, comicsList.get(eventBusAdapterClick.getPosition()));
            startActivity(intent);
        }
    }
}

File: app/src/main/java/com/codebosses/comical/ui/main/search/FragmentSearch.kt
package com.codebosses.comical.ui.main.search

import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.codebosses.comical.R
import com.codebosses.comical.common.Constants
import com.codebosses.comical.di.base.Injectable
import com.codebosses.comical.repository.eventbus.EventBusSearchClick
import com.codebosses.comical.repository.model.comics.ComicResult
import com.codebosses.comical.ui.detail.ComicDetailActivity
import com.codebosses.comical.ui.main.base.BaseFragment
import com.codebosses.comical.utils.ToastUtil
import com.codebosses.comical.utils.extensions.gone
import com.codebosses.comical.utils.extensions.observe
import com.codebosses.comical.utils.extensions.startActivity
import com.codebosses.comical.utils.extensions.visible
import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.fragment_search.*
import kotlinx.android.synthetic.main.fragment_search.view.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import javax.inject.Inject


class FragmentSearch : BaseFragment(), Injectable, TextWatcher {

    companion object {
        fun getInstance() = FragmentSearch()
    }

    //    View model fields....
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    lateinit var searchViewModel: SearchViewModel

    var delay: Long = 1000 // 1 seconds after user stops typing
    var last_text_edit: Long = 0
    var handler: Handler = Handler()

    //    Instance fields....
    private var chapterList = ArrayList<ComicResult>()
    private lateinit var searchComicAdapter: SearchComicAdapter

    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_search, container, false)
        EventBus.getDefault().register(this)

        searchViewModel = ViewModelProviders.of(this, viewModelFactory).get(SearchViewModel::class.java)

        with(view.recyclerViewSearch) {
            layoutManager = LinearLayoutManager(activity!!)
            searchComicAdapter = SearchComicAdapter(activity!!, chapterList)
            adapter = searchComicAdapter
        }

//        Listeners....
        view.editTextSearch.addTextChangedListener(this)
        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()
        EventBus.getDefault().unregister(this)
    }

    private val input_finish_checker = Runnable {
        if (System.currentTimeMillis() > last_text_edit + delay - 500) { // TODO: do what you need here
            searchComic(editTextSearch.text.toString())
        }
    }

    override fun afterTextChanged(s: Editable?) {
        if (s!!.isNotEmpty()) {
            last_text_edit = System.currentTimeMillis()
            handler.postDelayed(input_finish_checker, delay)
        }
    }

    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

    }

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        handler.removeCallbacks(input_finish_checker)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        AndroidSupportInjection.inject(this)
    }

    private fun searchComic(searchText: String) {
        searchViewModel.searchComic(searchText).observe(this) {
            when {
                it.status.isLoading() -> {
                    circularProgressBarSearch.visible()
                }
                it.status.isError() -> {
                    circularProgressBarSearch.gone()
                    ToastUtil.showCustomToast(activity!!, it.errorMessage!!)
                }
                it.status.isSuccessful() -> {
                    circularProgressBarSearch.gone()
                    searchComicAdapter.notifyItemRangeRemoved(0, chapterList.count())
                    chapterList.clear()
                    chapterList.addAll(it.data!!.result)
                    searchComicAdapter.notifyItemRangeInserted(0, chapterList.count())
                }
            }
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun eventBusSearchClick(eventBusSearchClick: EventBusSearchClick) {
        val bundle = Bundle()
        bundle.putParcelable(
                Constants.IntentConstants.COMIC_CHAPTER,
                chapterList[eventBusSearchClick.position]
        )
        activity!!.startActivity(ComicDetailActivity::class.java, bundle)
    }

}
